content/posts/2023-11-20-tools-for-learning-llvm-tablegen.md

```diff
@@ -0,0 +1,500 @@
+
+**Note:** You do not need to build a project to include its TableGen files in
+the compilation database. Configuring is all that is needed.
+
+Next, install the MLIR
+[extension](https://marketplace.visualstudio.com/items?itemName=llvm-vs-code-extensions.vscode-mlir)
+into Visual Studio Code (the language server works with other editors but their
+setup will be different).
+
+Then follow the setup instructions
+[here](https://mlir.llvm.org/docs/Tools/MLIRLSP/#td---tablegen-files) to tell
+the extension where the server and compilation database are.
+
+## Example
+
+This example assumes you have configured LLVM with the `AArch64` target enabled.
+(It is enabled by default)
+
+* Open the file `llvm/lib/Target/AArch64/AArch64.td`.
+* Put your cursor on a use of the `SubtargetFeature` type.
+* In the menu bar, select "Go" then "Go to Definition".
+* This takes you to `llvm/include/llvm/Target/Target.td`, where
+  `SubtargetFeature` is defined.
+
+## Limitations
+
+The language server highlights an anti-pattern in the way some LLVM targets
+such as AArch64 use TableGen.
+
+You may find yourself in a file that uses a class but does not define it or
+include any files which define it. This is because this file is intended to be
+included in another file, which does include a definition of that class.
+
+```
+example.td:
+  class Example {}
+
+uses_example.td:
+  def example: Example {}
+
+main.td:
+  include "example.td"
+  include "uses_example.td"
+```
+
+The example above shows this anti-pattern:
+
+* The file `example.td` defines the class `Example`.
+* `uses_example.td` uses the class `Example`, but does not include `example.td`.
+* `main.td` includes both `example.td` and `uses_example.td`.
+* `main.td` is the file that is compiled.
+* When you are in `uses_example.td`, the language server does not know where
+  `Example` is defined,
+* When you are in `main.td`, the language server does know where `Example` is
+  defined.
+
+Perhaps we can address this by improving the language server, or reorganising
+the includes so we do not have files that appear to be isolated.
+
+# Dump
+
+What about `printf`? The best debugging tool of them all.
+
+TableGen's equivalent is
+[dump](https://llvm.org/docs/TableGen/ProgRef.html#dump-print-messages-to-stderr),
+and its companion `repr`.
+```
+def op;
+class A {
+  string A = "some text";
+  dag X =(op op);
+}
+def a : A;
+
+dump "The Value of a is: \n" # !repr(a);
+```
+
+`dump` prints to `stderr`:
+```
+<source>:8:1: note: The Value of a is:
+a {	// A
+  string A = "some text";
+  dag X = (op op);
+}
+
+dump "The Value of a is: \n" # !repr(a);
+^
+```
+
+This was added
+[recently](https://github.com/llvm/llvm-project/commit/411c4edeef076bd2e01b104fe095ba381600a3d3)
+by [Francesco Petrogalli](https://github.com/fpetrogalli) and
+[Adam Nemet](https://github.com/anemet). So you will need a recent build, or a
+released version 18.0 or newer (which is unreleased at time of writing).
```

Comment:
I was thinking that being mentioned here is not important. The reference to the commit is already more than enough. 

```suggestion
This feature was added
[recently](https://github.com/llvm/llvm-project/commit/411c4edeef076bd2e01b104fe095ba381600a3d3).
You will need a recent build, or a
released version 18.0 or newer (which is unreleased at time of writing).
```

Reply:
Done.